### DISCUSSION TITLE
Problems with Berry introduction

### DISCUSSION THREAD

**Discussion by hagenbuch** [arendst/Tasmota#19604] [Q&A] [answered]
Hi all,

I'm not new to Tasmota but totally new to Berry.

I managed to flash an ESP32C3 firmware to this device I bought vie Tindie:

https://www.tindie.com/products/thehognl/esp32-c3-with-rs485-modbus-and-optional-touch-tft/

The device offers a "Berry console". Following this

https://tasmota.github.io/docs/Berry-Introduction/

I type in all variants of print("Hello"), with or without preceding > but all that happens after the blue "Run code" button is that my text is added to the orange text in the window above, beginning with "Welcome to the Berry console".

Also I get no response by typing 3 + 2 as stated in the documentation.

I tried it with Firefox 118 (on Linux Mint) but got the same result with Chrome
Tasmota version is 13.1.0.3

Where did I go wrong?

Would the standard firmware be already able to handle something like Modbus?

I'm happy that 2 M of the flash is being offered as "file system" so I guess I should be able to add (or write) converters etc.

Which binary would I need to also get the TFT ILI9341 running?

Andreas

**Comment by sfromis** [comment]
Not sure exactly what you see. On ESP32-C3 with Tasmota 13.1.0.3 I have:
![image](https://github.com/arendst/Tasmota/assets/47082390/0a5fc3ba-ba55-4843-84b0-e585a34e995a)

For ESP32-C3 and displays, I'd use [tasmota32c3-lvgl.bin](https://github.com/tasmota/install/blob/firmware/firmware/unofficial/tasmota32c3-lvgl.bin) with HASPmota/LVGL and [universal display](https://tasmota.github.io/docs/Displays/#universal-display-driver) with the appropriate display descriptor and pin configuration. This binary does not provide a "serial" console via USB CDC, but that may not be much of an issue.

**Reply by s-hadinger** [reply]
One possibility is that the device does a bootloop crash. At 4th boot, Berry is disabled although the console is still there but doing nothing.

Please check the boot logs, it says explicitly that Berry is not enabled

**Comment by hagenbuch** [comment, marked as answer]
@sfromis I don't have the white number and text, rest looks alike

@s-hadinger I tried a normal warm restart and it worked! Fascinating, thanks!

Log (console) says

`00:00:00.001 HDW: ESP32-C3 
00:00:00.013 UFS: FlashFS mounted with 176 kB free
00:00:00.023 CFG: Loaded from File, Count 13
00:00:00.029 QPC: Count 1
00:00:00.075 BRY: Berry initialized, RAM used=4382 bytes
00:00:00.085 TFS: File '.drvset003' not found
00:00:00.089 Project tasmota - Tasmota Version 13.1.0.3(tasmota32_2M)-2_0_13(2023-09-25T21:55:18)
00:00:00.622 WIF: Connecting to AP1 <secret> Channel 6 BSSId <secret> in mode 11n as tasmota-C11111-5984...
00:00:02.315 WIF: Connected
00:00:02.571 HTP: Web server active on tasmota-<secret>-5984 with IP address 192.168.13.86
15:08:19.559 RSL: INFO1 = {"Info1":{"Module":"ESP32C3","Version":"13.1.0.3(tasmota32_2M)","FallbackTopic":"cmnd/DVES_C11111_fb/","GroupTopic":"cmnd/tasmotas/"}}
15:08:19.571 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-C11111-5984","IPAddress":"192.168.13.86","IP6Global":"","IP6Local":"fe80::<secret>"}}
15:08:19.584 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset digital core","BootCount":7}}
15:08:22.967 QPC: Reset
15:08:23.974 RSL: STATE = {"Time":"2023-09-27T15:08:23","Uptime":"0T00:00:08","UptimeSec":8,"Heap":194,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":37},"Wifi":{"AP":1,"SSId":"<secret>","BSSId":"<secret>","Channel":6,"Mode":"11n","RSSI":70,"Signal":-65,"LinkCount":1,"Downtime":"0T00:00:03"}}`

**Reply by sfromis** [reply]
Beware that using a 2MB build means that you cannot do OTA upgrade of Tasmota, and some Tasmota features are not included. As a starting point, I'd always use a 4MB build including safeboot, and only consider a 2MB build on a 4MB device if having a very specific use case for extra file system capacity. You can still have a lot of berry code in a much smaller file system.

**Reply by hagenbuch** [reply]
Ah OK thank you.. so I guess I will have to compile my own then.

BTW: Is the ability to read and parse Modbus etc. already included in the standard build?

**Reply by sfromis** [reply]
What I wrote does not imply any reason to compile your own. Standard precompiled binaries for ESP32-C3 are for 4MB of flash; you are using a specialized outlier build to facilitate even using otherwise unworkable chips only having 2MB of flash in total. You also have the `tasmota32c3cdc` build if you want USB serial console support, but not if you also want the display features with LVGL builds.

Reading https://github.com/arendst/Tasmota/blob/development/BUILDS.md suggests that some Modbus support ( `USE_MODBUS_BRIDGE`) is included in the base build, but not having worked with Modbus, I'm not sure of the specifics. Anyway, using the tool [decode-status](https://github.com/tasmota/decode-status/), you can check if the binary you're using includes the feature.

**Reply by hagenbuch** [reply]
Great, again many thanks!
